### DISCUSSION TITLE
Support for logical operators in expressions

### DISCUSSION THREAD

**Discussion by zacdezgeo** [cogeotiff/rio-tiler#374] [Q&A] [answered]
Hello,

I was wondering if there was any detailed documentation about the supported operators for the expression query parameter. I found [this interesting article](https://observablehq.com/@vincentsarago/the-power-of-rio-tiler-expression) that points to the use of [numexpr](https://github.com/pydata/numexpr). 

More specifically I was wondering if logical operators were supported. I have been trying to have the following expression evaluated: 

`(B1>100&B1<150)*B1` 

After urlencoding:

`(B1%3E100%26B1%3C150)*B1`

Have there been any related discussions or are there any more docs on the issue?

Thanks for your help

**Comment by zacdezgeo** [comment, marked as answer]
Found the solution for logical operators, every condition must be encapsulated by parentheses: `((B1>50)%26(B1<150))*B1`

Are there any other cool expression features commonly used? Is there some way to integrate non pixel by pixel expressions? Example: calculating the mean of direct neighbors (3x3 window)

**Reply by vincentsarago** [reply]
@zacharyDez glad you found out the fix.

Usually I use expression just to get simple indexes like NDVI but doing more complex operations is def under my radar. I dream about being about to do the same a sentinel-hub ;-) but for now we are limited to what numexpr can do for us. 

>  Example: calculating the mean of direct neighbors (3x3 window)

for this you could use the `post_process` option (https://github.com/cogeotiff/rio-tiler/blob/master/rio_tiler/reader.py#L35-L37, https://github.com/cogeotiff/rio-tiler-pds/blob/master/rio_tiler_pds/landsat/aws/landsat8.py#L73-L80)

**Reply by zacdezgeo** [reply]
Thanks for pointing me in the right direction!
